### ISSUE TITLE
question on wh_iou

### ISSUE THREAD

**Issue by Chida15** [eriklindernoren/PyTorch-YOLOv3#270]
https://github.com/eriklindernoren/PyTorch-YOLOv3/blob/47b7c912877ca69db35b8af3a38d6522681b3bb3/utils/utils.py#L184

why the intersection area is calculated in this way.
![Screenshot from 2019-08-08 17-04-54](https://user-images.githubusercontent.com/39668508/62690124-af932a80-b9fe-11e9-8d7b-208cbead885c.png)
if two boxes are like this, the equation just doesn't work, I think.

**Comment by xiayanqian** [comment]
`ious = torch.stack([bbox_wh_iou(anchor, gwh) for anchor in anchors])
The function bbox_wh_iou is used to find the anchor which is most similar to ground truth box.
`inter_area = torch.min(w1, w2) * torch.min(h1, h2)`
the inter_area means the biggest area that can be reached when two rectangles intersect。
When the inter_area of anchors and groundtruth is the same,how to choose the best anchor?The union_area smaller,the anchor will be better.
So,the result of bbox_wh_iou shows the similarity between anchors and ground truth.

**Comment by whyz070205** [comment]
The centers of anchors and ground truth are the same cell. 

> https://github.com/eriklindernoren/PyTorch-YOLOv3/blob/47b7c912877ca69db35b8af3a38d6522681b3bb3/utils/utils.py#L184
> 
> why the intersection area is calculated in this way.
> ![Screenshot from 2019-08-08 17-04-54](https://user-images.githubusercontent.com/39668508/62690124-af932a80-b9fe-11e9-8d7b-208cbead885c.png)
> if two boxes are like this, the equation just doesn't work, I think.

It is computing on one cell in the feature map. So the centers of anchors and the ground truth are the same.

**Comment by ujsyehao** [comment]
@whyz070205 you are right.
ex:
each yolo layer has 3 anchor box
gt box number: 10
so the iou  calculate:
3 x 10 = 30
it has 30 iou scores.

**Comment by GuanRainy** [comment]
I agree @Chida15.I think the following is correct.
`
def bbox_wh_iou(anchor_wh:torch.Tensor, target_boxes:torch.Tensor):

    """
    :param anchor_wh: shape: (2, 2)
    :param target_boxes: shape: (#n, 4), row dim stand for (cx, cy, w, h)
    :return:
    """
    gxy_grid = target_boxes[:, :2].floor()
    anchor_wh = anchor_wh.view(1, 2).expand(gxy_grid.size(0), 2)
    anchor = torch.cat([gxy_grid, anchor_wh], dim=1)
    bboxs = xywh2xyxy(target_boxes)
    anchor = xywh2xyxy(anchor)
    x0 = torch.max(bboxs[:, 0], anchor[:, 0])
    y0 = torch.max(bboxs[:, 1], anchor[:, 1])
    x1 = torch.min(bboxs[:, 2], anchor[:, 2])
    y1 = torch.min(bboxs[:, 3], anchor[:, 3])
    inter_area = torch.clamp(x1 - x0, min=0)*torch.clamp(y1 - y0, min=0)
    bboxs_area = (bboxs[:, 2] - bboxs[:, 0])*(bboxs[:, 3] - bboxs[:, 1])
    anchor_area = (anchor[:, 2] - anchor[:, 0]) * (anchor[:, 3] - anchor[:, 1])
    iou = inter_area / (bboxs_area + anchor_area + 1e-8)
    return iou`
